File: ios/CustomMessageHandlerModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(CustomMessageHandlerModule, CustomMessageHandlerModule, NSObject)

RCT_EXTERN_METHOD(destroy:(NSString *)nativeId)

RCT_EXTERN__BLOCKING_SYNCHRONOUS_METHOD(onReceivedSynchronousMessageResult:(NSString *)nativeId result:(nullable NSString *)result)
RCT_EXTERN_METHOD(sendMessage:(NSString *)nativeId message:(NSString *)message data:(nullable NSString *)data)
RCT_EXTERN_METHOD(registerHandler:(NSString *)nativeId)
@end

File: src/ui/customMessageHandlerModule.ts
import { NativeModule, requireNativeModule } from 'expo-modules-core';

export type CustomMessageHandlerModuleEvents = {
  onReceivedSynchronousMessage: ({
    nativeId,
    id,
    message,
    data,
  }: {
    nativeId: string;
    id: number;
    message: string;
    data: string | undefined;
  }) => void;
  onReceivedAsynchronousMessage: ({
    nativeId,
    message,
    data,
  }: {
    nativeId: string;
    message: string;
    data: string | undefined;
  }) => void;
};

/**
 * Native CustomMessageHandlerModule using Expo modules API.
 * Provides modern async/await interface while maintaining backward compatibility.
 */
declare class CustomMessageHandlerModule extends NativeModule<CustomMessageHandlerModuleEvents> {
  registerHandler(nativeId: string): Promise<void>;
  destroy(nativeId: string): Promise<void>;
  onReceivedSynchronousMessageResult(
    id: number,
    result: string | undefined
  ): Promise<void>;
  sendMessage(
    nativeId: string,
    message: string,
    data: string | undefined
  ): Promise<void>;
}

export default requireNativeModule<CustomMessageHandlerModule>(
  'CustomMessageHandlerModule'
);
